File: udon_decompiler/analysis/transform/passes/collect_label_usage.py
from __future__ import annotations

from typing import Optional

from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRBlockContainer,
    IRContainerKind,
    IRFunction,
    IRHighLevelSwitch,
    IRIf,
    IRJump,
    IRLeave,
    IRReturn,
    IRStatement,
)
from udon_decompiler.analysis.transform.pass_base import (
    IILTransform,
    ILTransformContext,
)


class CollectLabelUsage(IILTransform):
    """
    Collect label usage info for codegen.

    After this pass:
    - IRBlock.should_emit_label marks whether `<label>:` must be emitted.
    - IRBlockContainer.should_emit_exit_label marks whether exit label must be emitted.
    """

    def run(self, function: IRFunction, context: ILTransformContext) -> None:
        _ = context
        self._reset_flags(function.body)
        self._analyze_container(
            container=function.body,
            break_target=None,
            function_body=function.body,
        )

    def _reset_flags(self, container: IRBlockContainer) -> None:
        container.should_emit_exit_label = False
        for block in container.blocks:
            block.should_emit_label = False
            for statement in block.statements:
                self._reset_in_statement(statement)

    def _reset_in_statement(self, statement: IRStatement) -> None:
        if isinstance(statement, IRIf):
            self._reset_in_statement(statement.true_statement)
            if statement.false_statement is not None:
                self._reset_in_statement(statement.false_statement)
            return

        if isinstance(statement, IRHighLevelSwitch):
            for section in statement.sections:
                section.body.should_emit_label = False
                for nested in section.body.statements:
                    self._reset_in_statement(nested)
            return

        if isinstance(statement, IRBlock):
            statement.should_emit_label = False
            for nested in statement.statements:
                self._reset_in_statement(nested)
            return

        if isinstance(statement, IRBlockContainer):
            self._reset_flags(statement)

    def _analyze_container(
        self,
        container: IRBlockContainer,
        break_target: Optional[IRBlockContainer],
        function_body: IRBlockContainer,
    ) -> None:
        if self._analyze_while_like(
            container=container,
            break_target=break_target,
            function_body=function_body,
        ):
            return

        if self._analyze_do_while_like(
            container=container,
            break_target=break_target,
            function_body=function_body,
        ):
            return

        for index, block in enumerate(container.blocks):
            next_block = (
                container.blocks[index + 1]
                if index + 1 < len(container.blocks)
                else None
            )
            for stmt_index, statement in enumerate(block.statements):
                statement_next_block = (
                    next_block if stmt_index == len(block.statements) - 1 else None
                )
                self._analyze_statement(
                    statement=statement,
                    current_container=container,
                    next_block=statement_next_block,
                    break_target=break_target,
                    function_body=function_body,
                )

    def _analyze_while_like(
        self,
        container: IRBlockContainer,
        break_target: Optional[IRBlockContainer],
        function_body: IRBlockContainer,
    ) -> bool:
        if container.kind != IRContainerKind.WHILE:
            return False

        entry = container.entry_block
        if entry is None or len(container.blocks) != 1 or not entry.statements:
            return False

        terminal = entry.statements[-1]
        if not isinstance(terminal, IRIf):
            return False
        if terminal.false_statement is None:
            return False
        if not self._is_leave_to_container(terminal.false_statement, container):
            return False
        if not isinstance(terminal.true_statement, IRBlock):
            return False

        body_statements = entry.statements[:-1] + terminal.true_statement.statements
        for statement in body_statements:
            self._analyze_statement(
                statement=statement,
                current_container=container,
                next_block=None,
                break_target=container,
                function_body=function_body,
            )
        return True

    def _analyze_do_while_like(
        self,
        container: IRBlockContainer,
        break_target: Optional[IRBlockContainer],
        function_body: IRBlockContainer,
    ) -> bool:
        _ = break_target
        if container.kind != IRContainerKind.DO_WHILE:
            return False

        entry = container.entry_block
        if entry is None or len(container.blocks) != 1 or not entry.statements:
            return False

        terminal = entry.statements[-1]
        if not isinstance(terminal, IRIf):
            return False
        if terminal.false_statement is None:
            return False
        if not self._is_leave_to_container(terminal.false_statement, container):
            return False
        if not self._is_branch_to_entry(terminal.true_statement, container):
            return False

        for statement in entry.statements[:-1]:
            self._analyze_statement(
                statement=statement,
                current_container=container,
                next_block=None,
                break_target=container,
                function_body=function_body,
            )
        return True

    def _analyze_statement(
        self,
        statement: IRStatement,
        current_container: Optional[IRBlockContainer],
        next_block: Optional[IRBlock],
        break_target: Optional[IRBlockContainer],
        function_body: IRBlockContainer,
    ) -> None:
        if isinstance(statement, IRIf):
            self._analyze_branch_statement(
                statement=statement.true_statement,
                current_container=current_container,
                break_target=break_target,
                function_body=function_body,
            )
            if statement.false_statement is not None:
                self._analyze_branch_statement(
                    statement=statement.false_statement,
                    current_container=current_container,
                    break_target=break_target,
                    function_body=function_body,
                )
            return

        if isinstance(statement, IRHighLevelSwitch):
            for section in statement.sections:
                for nested in section.body.statements:
                    self._analyze_statement(
                        statement=nested,
                        current_container=current_container,
                        next_block=None,
                        break_target=None,
                        function_body=function_body,
                    )
            return

        if isinstance(statement, IRJump):
            if (
                break_target is not None
                and break_target.entry_block is not None
                and statement.target is break_target.entry_block
            ):
                return
            if next_block is not None and statement.target is next_block:
                return
            statement.target.should_emit_label = True
            return

        if isinstance(statement, IRLeave):
            if statement.target_container is function_body:
                return
            if break_target is not None and statement.target_container is break_target:
                return
            statement.target_container.should_emit_exit_label = True
            return

        if isinstance(statement, IRReturn):
            return

        if isinstance(statement, IRBlock):
            for nested in statement.statements:
                self._analyze_statement(
                    statement=nested,
                    current_container=current_container,
                    next_block=None,
                    break_target=break_target,
                    function_body=function_body,
                )
            return

        if isinstance(statement, IRBlockContainer):
            self._analyze_container(
                container=statement,
                break_target=break_target,
                function_body=function_body,
            )

    def _analyze_branch_statement(
        self,
        statement: IRStatement,
        current_container: Optional[IRBlockContainer],
        break_target: Optional[IRBlockContainer],
        function_body: IRBlockContainer,
    ) -> None:
        if (
            isinstance(statement, IRBlock)
            and current_container is not None
            and statement in current_container.blocks
        ):
            statement.should_emit_label = True
            return

        self._analyze_statement(
            statement=statement,
            current_container=current_container,
            next_block=None,
            break_target=break_target,
            function_body=function_body,
        )

    @staticmethod
    def _is_leave_to_container(
        statement: IRStatement,
        container: IRBlockContainer,
    ) -> bool:
        return (
            isinstance(statement, IRLeave)
            and statement.target_container is container
        )

    @staticmethod
    def _is_branch_to_entry(
        statement: IRStatement,
        container: IRBlockContainer,
    ) -> bool:
        entry = container.entry_block
        if entry is None:
            return False

        if isinstance(statement, IRJump):
            return statement.target is entry

        if isinstance(statement, IRBlock) and len(statement.statements) == 1:
            nested = statement.statements[0]
            return isinstance(nested, IRJump) and nested.target is entry

        return False

File: src/decompiler/transform/passes/collect_label_usage.rs
use std::collections::HashSet;

use crate::decompiler::Result;
use crate::decompiler::ir::{
    IrBlockContainer, IrContainerKind, IrFunction, IrHighLevelDoWhile, IrHighLevelSwitch,
    IrHighLevelWhile, IrIf, IrJump, IrLeave, IrReturn, IrStatement,
};
use crate::decompiler::transform::pass_base::{ITransform, TransformContext};

pub struct CollectLabelUsage;

impl ITransform for CollectLabelUsage {

    fn run(&self, function: &mut IrFunction, _context: &mut TransformContext<'_, '_>) -> Result<()> {
        reset_flags(&mut function.body);

        let mut label_targets = HashSet::<u32>::new();
        let mut exit_label_targets = HashSet::<u32>::new();

        analyze_container(
            &function.body,
            None,
            None,
            function.body.id,
            &mut label_targets,
            &mut exit_label_targets,
        );

        apply_flags(&mut function.body, &label_targets, &exit_label_targets);

        Ok(())
    }
}

fn reset_flags(container: &mut IrBlockContainer) {
    container.should_emit_exit_label = false;
    for block in &mut container.blocks {
        block.should_emit_label = false;
        for statement in &mut block.statements {
            reset_statement_flags(statement);
        }
    }
}

fn reset_statement_flags(statement: &mut IrStatement) {
    match statement {
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            reset_statement_flags(true_statement);
            if let Some(false_statement) = false_statement.as_mut() {
                reset_statement_flags(false_statement);
            }
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                section.body.should_emit_label = false;
                for nested in &mut section.body.statements {
                    reset_statement_flags(nested);
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => reset_flags(body),
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => reset_flags(body),
        IrStatement::Block(block) => {
            block.should_emit_label = false;
            for nested in &mut block.statements {
                reset_statement_flags(nested);
            }
        }
        IrStatement::BlockContainer(container) => reset_flags(container),
        _ => {}
    }
}

fn analyze_container(
    container: &IrBlockContainer,
    break_target: Option<u32>,
    continue_target: Option<u32>,
    function_body_id: u32,
    label_targets: &mut HashSet<u32>,
    exit_label_targets: &mut HashSet<u32>,
) {
    if analyze_while_like(
        container,
        break_target,
        continue_target,
        function_body_id,
        label_targets,
        exit_label_targets,
    ) {
        return;
    }

    if analyze_do_while_like(
        container,
        break_target,
        continue_target,
        function_body_id,
        label_targets,
        exit_label_targets,
    ) {
        return;
    }

    for index in 0..container.blocks.len() {
        let block = &container.blocks[index];
        let next_block = container.blocks.get(index + 1).map(|x| x.start_address);

        for stmt_index in 0..block.statements.len() {
            let statement = &block.statements[stmt_index];
            let statement_next_block = if stmt_index + 1 == block.statements.len() {
                next_block
            } else {
                None
            };
            analyze_statement(
                statement,
                container,
                statement_next_block,
                break_target,
                continue_target,
                function_body_id,
                label_targets,
                exit_label_targets,
            );
        }
    }
}

fn analyze_while_like(
    container: &IrBlockContainer,
    _break_target: Option<u32>,
    continue_target: Option<u32>,
    function_body_id: u32,
    label_targets: &mut HashSet<u32>,
    exit_label_targets: &mut HashSet<u32>,
) -> bool {
    if container.kind != IrContainerKind::While {
        return false;
    }

    let Some(entry) = container.entry_block() else {
        return false;
    };
    if container.blocks.len() != 1 || entry.statements.is_empty() {
        return false;
    }

    let Some(IrStatement::If(IrIf {
        true_statement,
        false_statement,
        ..
    })) = entry.statements.last()
    else {
        return false;
    };
    let Some(false_statement) = false_statement.as_ref() else {
        return false;
    };
    if !is_leave_to_container(false_statement, container.id) {
        return false;
    }

    let IrStatement::Block(true_block) = true_statement.as_ref() else {
        return false;
    };

    let mut body_statements = entry.statements[..entry.statements.len() - 1].to_vec();
    body_statements.extend(true_block.statements.clone());

    for statement in &body_statements {
        analyze_statement(
            statement,
            container,
            None,
            Some(container.id),
            continue_target,
            function_body_id,
            label_targets,
            exit_label_targets,
        );
    }

    true
}

fn analyze_do_while_like(
    container: &IrBlockContainer,
    _break_target: Option<u32>,
    continue_target: Option<u32>,
    function_body_id: u32,
    label_targets: &mut HashSet<u32>,
    exit_label_targets: &mut HashSet<u32>,
) -> bool {
    if container.kind != IrContainerKind::DoWhile {
        return false;
    }

    let Some(entry) = container.entry_block() else {
        return false;
    };
    if container.blocks.len() != 1 || entry.statements.is_empty() {
        return false;
    }

    let Some(IrStatement::If(IrIf {
        true_statement,
        false_statement,
        ..
    })) = entry.statements.last()
    else {
        return false;
    };
    let Some(false_statement) = false_statement.as_ref() else {
        return false;
    };
    if !is_leave_to_container(false_statement, container.id) {
        return false;
    }
    if !is_branch_to_entry(
        true_statement,
        container.entry_block().map(|x| x.start_address),
    ) {
        return false;
    }

    for statement in &entry.statements[..entry.statements.len() - 1] {
        analyze_statement(
            statement,
            container,
            None,
            Some(container.id),
            continue_target,
            function_body_id,
            label_targets,
            exit_label_targets,
        );
    }

    true
}

#[allow(clippy::too_many_arguments)]
fn analyze_statement(
    statement: &IrStatement,
    current_container: &IrBlockContainer,
    next_block: Option<u32>,
    break_target: Option<u32>,
    continue_target: Option<u32>,
    function_body_id: u32,
    label_targets: &mut HashSet<u32>,
    exit_label_targets: &mut HashSet<u32>,
) {
    match statement {
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            analyze_branch_statement(
                true_statement,
                current_container,
                break_target,
                continue_target,
                function_body_id,
                label_targets,
                exit_label_targets,
            );
            if let Some(false_statement) = false_statement.as_ref() {
                analyze_branch_statement(
                    false_statement,
                    current_container,
                    break_target,
                    continue_target,
                    function_body_id,
                    label_targets,
                    exit_label_targets,
                );
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile {
            body,
            break_target: while_break,
            continue_target: while_continue,
            ..
        }) => {
            analyze_container(
                body,
                Some(*while_break),
                Some(*while_continue),
                function_body_id,
                label_targets,
                exit_label_targets,
            );
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile {
            body,
            break_target: loop_break,
            continue_target: loop_continue,
            ..
        }) => {
            analyze_container(
                body,
                Some(*loop_break),
                Some(*loop_continue),
                function_body_id,
                label_targets,
                exit_label_targets,
            );
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                for nested in &section.body.statements {
                    analyze_statement(
                        nested,
                        current_container,
                        None,
                        None,
                        None,
                        function_body_id,
                        label_targets,
                        exit_label_targets,
                    );
                }
            }
        }
        IrStatement::Jump(IrJump { target_address }) => {
            if continue_target.is_some_and(|x| x == *target_address) {
                return;
            }
            if next_block.is_some_and(|x| x == *target_address) {
                return;
            }
            if !current_container
                .blocks
                .iter()
                .any(|block| block.start_address == *target_address)
            {
                return;
            }
            label_targets.insert(*target_address);
        }
        IrStatement::Leave(IrLeave {
            target_container_id,
        }) => {
            if *target_container_id == function_body_id {
                return;
            }
            if break_target.is_some_and(|x| x == *target_container_id) {
                return;
            }
            exit_label_targets.insert(*target_container_id);
        }
        IrStatement::Return(IrReturn) => {}
        IrStatement::Block(block) => {
            for nested in &block.statements {
                analyze_statement(
                    nested,
                    current_container,
                    None,
                    break_target,
                    continue_target,
                    function_body_id,
                    label_targets,
                    exit_label_targets,
                );
            }
        }
        IrStatement::BlockContainer(container) => {
            analyze_container(
                container,
                break_target,
                continue_target,
                function_body_id,
                label_targets,
                exit_label_targets,
            );
        }
        _ => {}
    }
}

fn analyze_branch_statement(
    statement: &IrStatement,
    current_container: &IrBlockContainer,
    break_target: Option<u32>,
    continue_target: Option<u32>,
    function_body_id: u32,
    label_targets: &mut HashSet<u32>,
    exit_label_targets: &mut HashSet<u32>,
) {
    if let IrStatement::Block(block) = statement
        && current_container
            .blocks
            .iter()
            .any(|x| x.start_address == block.start_address)
    {
        label_targets.insert(block.start_address);
        return;
    }

    analyze_statement(
        statement,
        current_container,
        None,
        break_target,
        continue_target,
        function_body_id,
        label_targets,
        exit_label_targets,
    );
}

fn apply_flags(
    container: &mut IrBlockContainer,
    label_targets: &HashSet<u32>,
    exit_label_targets: &HashSet<u32>,
) {
    container.should_emit_exit_label = exit_label_targets.contains(&container.id);
    for block in &mut container.blocks {
        block.should_emit_label = label_targets.contains(&block.start_address);
        for statement in &mut block.statements {
            apply_flags_statement(statement, label_targets, exit_label_targets);
        }
    }
}

fn apply_flags_statement(
    statement: &mut IrStatement,
    label_targets: &HashSet<u32>,
    exit_label_targets: &HashSet<u32>,
) {
    match statement {
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            apply_flags_statement(true_statement, label_targets, exit_label_targets);
            if let Some(false_statement) = false_statement.as_mut() {
                apply_flags_statement(false_statement, label_targets, exit_label_targets);
            }
        }
        IrStatement::Block(block) => {
            block.should_emit_label = label_targets.contains(&block.start_address);
            for nested in &mut block.statements {
                apply_flags_statement(nested, label_targets, exit_label_targets);
            }
        }
        IrStatement::BlockContainer(container) => {
            apply_flags(container, label_targets, exit_label_targets)
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                section.body.should_emit_label =
                    label_targets.contains(&section.body.start_address);
                for nested in &mut section.body.statements {
                    apply_flags_statement(nested, label_targets, exit_label_targets);
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            apply_flags(body, label_targets, exit_label_targets)
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            apply_flags(body, label_targets, exit_label_targets)
        }
        _ => {}
    }
}

fn is_leave_to_container(statement: &IrStatement, container_id: u32) -> bool {
    matches!(
        statement,
        IrStatement::Leave(IrLeave { target_container_id }) if *target_container_id == container_id
    )
}

fn is_branch_to_entry(statement: &IrStatement, entry_address: Option<u32>) -> bool {
    let Some(entry_address) = entry_address else {
        return false;
    };

    match statement {
        IrStatement::Jump(IrJump { target_address }) => *target_address == entry_address,
        IrStatement::Block(block) if block.statements.len() == 1 => matches!(
            block.statements.first(),
            Some(IrStatement::Jump(IrJump { target_address })) if *target_address == entry_address
        ),
        _ => false,
    }
}
